### PR TITLE
Update VisualElement docs

### PR DESCRIPTION
**Pull request by jfversluis**
### Description of Change

Moves the external comments to inline, adds missing comments and minor improvements.

### Issues Fixed

Related to https://github.com/dotnet/maui/issues/3960

### CODE REVIEW

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 1]
```diff
@@ -12,13 +12,18 @@
 
 namespace Microsoft.Maui.Controls
 {
-	/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="Type[@FullName='Microsoft.Maui.Controls.VisualElement']/Docs/*" />
+	/// <summary>
+	/// A <see cref="Element" /> that occupies an area on the screen, has a visual appearance, and can obtain touch input.
+	/// </summary>
+	/// <remarks>
+	/// The base class for most Microsoft.Maui.Controls on-screen elements. Provides most properties, events, and methods for presenting an item on screen.
+    /// </remarks>
```
Comment (rmarinho) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:20]: formatting

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 2]
```diff
@@ -864,10 +1019,17 @@ internal LayoutConstraint SelfConstraint
 			}
 		}
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='BatchBegin']/Docs/*" />
+		/// <summary>
+		/// Signals the start of a batch of changes to the elements properties. This can benefit performance if a bigger number of property values are changed.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1023]: I'd move the second sentence to the remarks

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 3]
```diff
@@ -864,10 +1019,17 @@ internal LayoutConstraint SelfConstraint
...
+		/// When the operation is done, <see cref="BatchCommit"/> should be called.
+		/// </summary>
+		/// <remarks>Application authors will not generally need to batch updates manually as the animation framework will do this for them.</remarks>
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1026]:
```suggestion
		/// <remarks>Application authors will generally not need to batch updates manually, as the animation framework will do this for them.</remarks>
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 4]
```diff
@@ -906,15 +1075,31 @@ public ResourceDictionary Resources
...
 
+		/// <summary>
+		/// Signals that the platform equivalent element for this element's size has changed and a new layout cycle might be needed.
+		/// </summary>
+		/// <remarks>For internal use only. This API can be changed or removed without notice at any time.</remarks>
 		[EditorBrowsable(EditorBrowsableState.Never)]
 		public void PlatformSizeChanged() => InvalidateMeasureInternal(InvalidationTrigger.MeasureChanged);
 
+		/// <summary>
+		/// Occurs when children of this element are changed in order.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1086]:
```suggestion
		/// Occurs when order of this element's children changes.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 5]
```diff
@@ -906,15 +1075,31 @@ public ResourceDictionary Resources
...
+		/// </summary>
 		public event EventHandler ChildrenReordered;
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='Focus']/Docs/*" />
+		/// <summary>
+		/// Attempts to set focus to this element.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1091]:
```suggestion
		/// Attempts to set the keyboard focus on this element.
```
I think we might need to be clearer that this focus is for entries or keyboard-focusable stuff. That being said I'm not even 100% sure what this property does so I might be wrong.

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 6]
```diff
@@ -968,7 +1153,14 @@ SizeRequest GetSizeRequest(double widthConstraint, double heightConstraint)
...
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='Measure']/Docs/*" />
+		/// <summary>
+		/// Returns the minimum size that a visual element needs in order to be displayed on the device.
+		/// </summary>
+		/// <param name="widthConstraint">The suggested maximum width constraint for the visual element to render.</param>
+		/// <param name="heightConstraint">The suggested maximum height constraint for the visual element to render.</param>
+		/// <param name="flags">A value that controls whether margins are included in the returned size.</param>
+		/// <returns>The minimum size that a visual element needs in order to be displayed on the device.</returns>
+		/// <remarks>If the minimum size that the visual element needs in order to be displayed on the device is larger than can be accommodated by<paramref name="widthConstraint" /> and <paramref name="heightConstraint" />, the return value may represent a rectangle that is larger in either one or both of those parameters.</remarks>
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1163]:
```suggestion
		/// <remarks>If the minimum size that the visual element needs in order to be displayed on the device is larger than can be accommodated by <paramref name="widthConstraint" /> and <paramref name="heightConstraint" />, the return value may represent a rectangle that is larger in either one or both of those parameters.</remarks>
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 7]
```diff
@@ -993,11 +1185,20 @@ public virtual SizeRequest Measure(double widthConstraint, double heightConstrai
...
 		public event EventHandler MeasureInvalidated;
 
+		/// <summary>
+		/// Occurs when the size of a <see cref="VisualElement"/> changed.
+		/// </summary>
 		public event EventHandler SizeChanged;
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='Unfocus']/Docs/*" />
+		/// <summary>
+		/// Unsets focus to this element.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1199]:
```suggestion
		/// Unsets keyboard focus on this element.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 8]
```diff
@@ -1038,11 +1246,20 @@ protected override void OnChildRemoved(Element child, int oldLogicalIndex)
 				view.ComputedConstraint = LayoutConstraint.None;
 		}
 
+		/// <summary>
+		/// Causes the <see cref="ChildrenReordered"/> event to be triggered.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1250]:
```suggestion
		/// Raises the <see cref="ChildrenReordered"/> event.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 9]
```diff
@@ -1038,11 +1246,20 @@ protected override void OnChildRemoved(Element child, int oldLogicalIndex)
...
 			=> ChildrenReordered?.Invoke(this, EventArgs.Empty);
 
 		IPlatformSizeService _platformSizeService;
 
+		/// <summary>
+		/// Method that is called when a layout measurement happens.
+		/// </summary>
+		/// <param name="widthConstraint">The width constraint to request.</param>
+		/// <param name="heightConstraint">The height constraint to request.</param>
+		/// <returns>The requested size that a visual element wants in order to be displayed on the device.</returns>
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1262]:
```suggestion
		/// <returns>The requested size that the element requires in order to be displayed on the device.</returns>
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 10]
```diff
@@ -1055,12 +1272,27 @@ protected virtual SizeRequest OnMeasure(double widthConstraint, double heightCon
 			return _platformSizeService.GetPlatformSize(this, widthConstraint, heightConstraint);
 		}
 
+		/// <summary>
+		/// This method is called when the size of the element is set during a layout cycle. Implement this method to add class handling for this event.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1276]:
```suggestion
		/// Method that is called when the size of the element is set during a layout cycle. Implement this method to add class handling for this event.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 11]
```diff
@@ -1055,12 +1272,27 @@ protected virtual SizeRequest OnMeasure(double widthConstraint, double heightCon
...
+		/// </summary>
+		/// <param name="width">The new width of the element.</param>
+		/// <param name="height">The new height of the element.</param>
 		protected virtual void OnSizeAllocated(double width, double height)
 		{
 		}
 
+		/// <summary>
+		/// This method is called during a layout cycle to signal the start of a sub-tree layout.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1285]:
```suggestion
		/// Method that is called during a layout cycle to signal the start of a sub-tree layout.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 12]
```diff
@@ -1519,14 +1766,25 @@ static void ZIndexPropertyChanged(BindableObject bindable, object oldValue, obje
...
 			set { SetValue(ZIndexProperty, value); }
 		}
 
+		/// <summary>
+		/// Gets the size that this <see cref="VisualElement"/> computed during the measure pass of the layout process.
+		/// </summary>
 		public Size DesiredSize { get; protected set; }
 
+		/// <summary>
+		/// Positions child objects and determines a size for a <see cref="VisualElement"/>. Parent objects that implement custom layout for their child elements should call this method from their layout override implementations to form a recursive layout update.
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1785]: I'd recommend putting the second sentence in the remarks

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 13]
```diff
@@ -1537,16 +1795,24 @@ Size IView.Arrange(Rect bounds)
 			return ArrangeOverride(bounds);
 		}
 
-		// ArrangeOverride provides a way to allow subclasses (e.g., ScrollView) to override Arrange even though
-		// the interface has to be explicitly implemented to avoid conflict with the old Arrange method
+		/// <summary>
+		/// Provides a way to allow subclasses (e.g., <see cref="ScrollView"/>) to override <see cref="Arrange(Rect)"/> even though
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1799]:
```suggestion
		/// Allows subclasses to override <see cref="Arrange(Rect)"/> even though
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 14]
```diff
@@ -1570,8 +1838,13 @@ Size IView.Measure(double widthConstraint, double heightConstraint)
 			return MeasureOverride(widthConstraint, heightConstraint);
 		}
 
-		// MeasureOverride provides a way to allow subclasses (e.g., Layout) to override Measure even though
-		// the interface has to be explicitly implemented to avoid conflict with the old Measure method
+		/// <summary>
+		/// MeasureOverride provides a way to allow subclasses (e.g., Layout) to override Measure even though
```
Comment (jknaudt21) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1842]:
```suggestion
		/// Provides a way to allow subclasses to override <see cref="Measure"/> even though
```

**Review by jknaudt21 — CHANGES_REQUESTED** [review]

**Comment by jknaudt21** [comment]
We also need to document the explicit interface implementations since the docs tooling will not automatically grab these. Hopefully an `<inheritdoc>` should be enough. 

A lot of them come from IView and IElement, and are properties we already have documented, such as `Height`. 

To get a quick reference of the items that are missing, see: https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.visualelement?view=net-maui-7.0#explicit-interface-implementations

File: src/Controls/src/Core/VisualElement/VisualElement.cs
[the remark this answers is not in the document]
Comment (jfversluis) [reply src/Controls/src/Core/VisualElement/VisualElement.cs:1]: Yeah I was on the fence about that one. Looking at other Microsoft frameworks they mix it as well, but you're right, let me add some uniformity here.

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 15]
```diff
@@ -880,7 +1049,14 @@ public void BatchCommit()
 		ResourceDictionary _resources;
 		bool IResourcesProvider.IsResourcesCreated => _resources != null;
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='Resources']/Docs/*" />
+		/// <summary>
+		/// Gets or sets the local resource dictionary.
+		/// </summary>
+		/// <remarks>
+		/// <para>In XAML, resource dictionaries are filled with key/value pairs that are specified in XAML and consequently created at run time. The keys in the resource dictionary are specified with the <c>x:Key</c> attribute of the XAML tag for the type to create. An object of that type is created, and is initialized with the property and field values that are specified either by additional attributes or by nested tags, both of which, when present are simply string representations of the property or field names. The object is then inserted into the <see cref="ResourceDictionary" /> for the enclosing type at runtime.</para>
```
Comment (samhouts) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1056]:
```suggestion
		/// <para>In XAML, resource dictionaries are filled with key/value pairs that are specified in XAML and consequently created at runtime. The keys in the resource dictionary are specified with the <c>x:Key</c> attribute of the XAML tag for the type to create. An object of that type is created, and is initialized with the property and field values that are specified either by additional attributes or by nested tags, both of which, when present, are simply string representations of the property or field names. The object is then inserted into the <see cref="ResourceDictionary" /> for the enclosing type at runtime.</para>
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 16]
```diff
@@ -880,7 +1049,14 @@ public void BatchCommit()
...
+		/// <para>Resource dictionaries and their associated XML provide the application developer with a convenient method to reuse code inside the XAML compile-time and run-time engines.</para>
```
Comment (samhouts) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1057]:
```suggestion
		/// <para>Resource dictionaries and their associated XML provide the application developer with a convenient method to reuse code inside the XAML compile-time and runtime engines.</para>
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 17]
```diff
@@ -1519,78 +1777,116 @@ static void ZIndexPropertyChanged(BindableObject bindable, object oldValue, obje
 			}
 		}
 
+		/// <summary>
+		/// Gets and sets the front-to-back z-index of an element within a layout. This is a bindable property.
```
Comment (samhouts) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1781]:
```suggestion
		/// Gets or sets the front-to-back z-index of an element within a layout. This is a bindable property.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 18]
```diff
@@ -1519,78 +1777,116 @@ static void ZIndexPropertyChanged(BindableObject bindable, object oldValue, obje
...
 		}
 
-		// ArrangeOverride provides a way to allow subclasses (e.g., ScrollView) to override Arrange even though
-		// the interface has to be explicitly implemented to avoid conflict with the old Arrange method
+		/// <summary>
+		/// Allows subclasses to override <see cref="Arrange(Rect)"/> even though
+		/// the interface has to be explicitly implemented to avoid conflict with the old <see cref="Arrange(Rect)"/> method.
+		/// </summary>
+		/// <param name="bounds">The new bounds of the element.</param>
+		/// <returns>The resulting size of this elements frame by the platform.</returns>
```
Comment (samhouts) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:1816]:
```suggestion
		/// <returns>The resulting size of this element's frame by the platform.</returns>
```

**Review by samhouts — CHANGES_REQUESTED** [review]

**Review by samhouts — APPROVED** [review]

**Review by jknaudt21 — APPROVED** [review]

**Comment by jfversluis** [comment]
Should be relatively easy to backport and will increase the coverage of the docs in vCurrent

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 19]
```diff
@@ -624,146 +676,240 @@ private protected bool InputTransparentCore
...
+		/// </para>
+		/// </remarks>
 		public double MinimumWidthRequest
 		{
 			get { return (double)GetValue(MinimumWidthRequestProperty); }
 			set { SetValue(MinimumWidthRequestProperty, value); }
 		}
 
+		/// <summary>
+		/// Gets or sets a value which overrides the maximum height the element will request during layout. This is a bindable property.
```
Comment (hartez) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:725]:
```suggestion
		/// Gets or sets the maximum height the element will request during layout. This is a bindable property.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 20]
```diff
@@ -624,146 +676,240 @@ private protected bool InputTransparentCore
...
+		/// </summary>
+		/// <remarks>
+		/// <para>The default value is <see cref="double.PositiveInfinity"/>.</para>
+		/// <para><see cref="MaximumHeightRequest"/> is used to override the results of a call to <see cref="GetSizeRequest(double, double)"/> by setting the maximum height property. This causes overflow handling to shrink this element to its maximum height before elements who do not have a maximum size set.
```
Comment (hartez) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:729]:
```suggestion
		/// <para><see cref="MaximumHeightRequest"/> is used to ensure that the element has no more than the specified height during layout.
```

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 21]
```diff
@@ -624,146 +676,240 @@ private protected bool InputTransparentCore
...
 		{
 			get { return _mockX == -1 ? (double)GetValue(XProperty) : _mockX; }
 			private set { SetValue(XPropertyKey, value); }
 		}
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/VisualElement.xml" path="//Member[@MemberName='Y']/Docs/*" />
+		/// <summary>
+		/// Gets the current Y position of this element. This is a read-only bindable property.
+		/// </summary>
+		/// <remarks>The y value of an element is set during the layout cycle.</remarks>
```
Comment (hartez) [inline src/Controls/src/Core/VisualElement/VisualElement.cs:891]:
```suggestion
		/// <remarks>The position of an element is set during layout.</remarks>
```

**Review by hartez — CHANGES_REQUESTED** [review]

**Pull request merged**